src/main/java/com/iota/iri/utils/dag/DAGUtils.java

```diff
@@ -0,0 +1,194 @@
+package com.iota.iri.utils.dag;
+
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.storage.Tangle;
+
+import java.util.*;
+
+/**
+ * This class offers generic functions for recurring tasks that are related to the tangle and that otherwise would have
+ * to be implemented over and over again in different parts of the code.
+ */
+public class DAGUtils {
+    /**
+     * Holds references to the singleton DAGUtils instances.
+     */
+    protected static HashMap<Tangle, DAGUtils> instances = new HashMap<>();
+
+    /**
+     * Holds a reference to the tangle instance which acts as an interface to the used database.
+     */
+    protected Tangle tangle;
+
+    /**
+     * This method allows us to retrieve the DAGUtils instance that corresponds to the given parameters.
+     *
+     * To save memory and to be able to use things like caches effectively, we do not allow to create multiple instances
+     * of this class for the same set of parameters, but create a singleton instance of it.
+     *
+     * @param tangle database interface used to retrieve data
+     * @return DAGUtils instance that allows to use the implemented methods.
+     */
+    public static DAGUtils get(Tangle tangle) {
+        DAGUtils instance = instances.get(tangle);
+        if(instance == null) {
```

Comment:
now notice that this isn't thread safe. Two different threads will call `get` the same time and each will return a different `DAGUtils` object.

I think that under your current code this will not cause any problems (there will just be a waste). However in the future some smart dev may come along and will add another field to DAGUtils…
We want to create a design that won't do anything surprising.

There are 2 ways to circumvent this:

1. Write safe thread code using `synchronized`. You can also look at [double-check locking](https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java), but this is ugly,
2. (my preferred way) - I usually think of a utils class as a collection of static methods. If you just have static methods that accept `tangle` as a parameter you will gain the following benefits:
  - None of this thread safety bs
  - Static will run faster than instance methods
  - You wont encourage other devs to add fields to this class (instances are good for OOP, but for utils I think this way is better)